drv/ksz8463/src/lib.rs

```diff
@@ -0,0 +1,149 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+#![no_std]
+
+use core::convert::TryInto;
+
+use drv_spi_api::{SpiDevice, SpiError};
+use drv_stm32h7_gpio_api as gpio_api;
+use gpio_api::{Gpio, OutputType, PinSet, Pull, Speed};
+use ringbuf::*;
+use userlib::{hl::sleep_for, task_slot};
+
+task_slot!(GPIO, gpio_driver);
+
+#[derive(Copy, Clone, Debug, PartialEq)]
+enum Trace {
+    None,
+    Read(Register, u16),
+    Write(Register, u16),
+    Id(u16),
+}
+ringbuf!(Trace, 16, Trace::None);
+
+const fn register_offset(address: u16) -> u16 {
+    let addr10_2 = address >> 2;
+    let mask_shift = 2 /* turn around bits */ + (2 * ((address >> 1) & 0x1));
+    (addr10_2 << 6) | ((0x3 as u16) << mask_shift)
+}
+
+#[derive(Copy, Clone, Debug, PartialEq)]
+#[repr(u16)]
+#[allow(non_camel_case_types)]
+pub enum Register {
+    CIDER = register_offset(0x000),
+    SGCR1 = register_offset(0x002),
+    SGCR2 = register_offset(0x004),
+    SGCR3 = register_offset(0x006),
+    SGCR6 = register_offset(0x00c),
+    SGCR7 = register_offset(0x00e),
+    MACAR1 = register_offset(0x010),
+    MACAR2 = register_offset(0x012),
+    MACAR3 = register_offset(0x014),
+
+    P1MBCR = register_offset(0x04c),
+    P1MBSR = register_offset(0x04e),
+
+    P2MBCR = register_offset(0x058),
+    P2MBSR = register_offset(0x05a),
+
+    CFGR = register_offset(0x0d8),
+    DSP_CNTRL_6 = register_offset(0x734),
+}
+
+pub struct Ksz8463 {
+    spi: SpiDevice,
+    nrst: PinSet,
+    slow_reset: bool,
+}
+
+impl Ksz8463 {
+    pub fn new(spi: SpiDevice, nrst: PinSet, slow_reset: bool) -> Self {
+        Self {
+            spi,
+            nrst,
+            slow_reset,
+        }
+    }
+
+    pub fn read(&self, r: Register) -> Result<u16, SpiError> {
+        let cmd = (r as u16).to_be_bytes();
+        let request = [cmd[0], cmd[1]];
+        let mut response = [0; 4];
+
+        self.spi.exchange(&request, &mut response)?;
+        let v = u16::from_le_bytes(response[2..].try_into().unwrap());
+        ringbuf_entry!(Trace::Read(r, v));
+
+        Ok(v)
+    }
+
+    pub fn write(&self, r: Register, v: u16) -> Result<(), SpiError> {
+        let cmd = (r as u16 | 0x8000).to_be_bytes(); // Set MSB to indicate write.
+        let data = v.to_le_bytes();
+        let request = [cmd[0], cmd[1], data[0], data[1]];
+
+        ringbuf_entry!(Trace::Write(r, v));
+        self.spi.write(&request[..])?;
+        Ok(())
+    }
+
+    pub fn write_masked(
+        &self,
+        r: Register,
+        v: u16,
+        mask: u16,
+    ) -> Result<(), SpiError> {
+        let _v = (self.read(r)? & !mask) | (v & mask);
```

Comment:
It's pretty deeply ingrained in Rust that variables starting with underscore are unused, so turning around and using this seems confusing.